travis: Fixup rpmbuild test to actually test the dependencies, add Debian testing, test distclean.

olifre opened the pull request:
**Description of the Pull Request (PR):**

This fixes the packaging test for RedHat systems, by actually testing the dependencies listed in the spec-file. 


**This fixes or addresses the following GitHub issues:**

- Ref: #1203 


**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

olifre committed:
travis: Fixup rpmbuild test to actually test the dependencies.

Leverage yum-builddep for that.

Signed-off-by: Oliver Freyermuth <o.freyermuth@googlemail.com>

olifre committed:
travis: Add testing and packaging test for Debian 8 and 9.

olifre committed:
Add test that distclean target works.

c-ney reviewed (commented):
I think `yum-builddep` doesn't pull `squashfs-tools` in because it is a `Requires` in the .spec file, not a `BuildRequires`. For runtime dependencies I suggest just using `yum install -y ~/rpmbuild/RPMS/*/*.rpm` instead of `rpm -iv`.

Edit: This should've been a comment/"review" on the specific commit only, guess the button doesn't work like that.

olifre committed:
travis: Specify --create-home when creating testuser.

On many distributions, creation of the home directory
is not handled by useradd by default.

jmstover commented on .travis/rpmbuild_test:
```diff
@@ -5,13 +5,13 @@
-yum install -y libtool rpm-build make yum-utils squashfs-tools
+yum install -y libtool rpm-build make yum-utils
 ./autogen.sh
 yum-builddep -y singularity.spec
```
This won't work... singularity.spec is not generated until after `./configure` runs.

If you specified `yum-builddep -y singularity.spec.in` would that grab the BuildRequires entries, or would it break because it isn't "complete"?

olifre replied:
Good point!
Any idea what's the magic trick why it still works in travis nevertheless? 
Here's the direct link to the build log:
https://api.travis-ci.org/v3/job/331238663/log.txt
Just search for `yum-builddep` and you will arrive at the place which ought not work...

jmstover replied:
Don't know why that succeeds. There was a build earlier, where `yum-builddep` blew up because `singularity.spec` didn't exist... and it actually _shouldn't_ exist at that point...

Where is the singularity source being grabbed from? I don't see a checkout or tarball extraction prior to the `./autogen.sh` run. A release tar extraction should work, as that would be packaged with a singularity.spec file...  But not if you're building directly from a clean git pull.

dtrudg replied:
@jmstover see `.travis/centos_run` the source is bound into the docker container from `pwd` where it will be in the travis build environment:

```
docker run --privileged -d -ti -e "container=docker" -v /var/run/docker.sock:/docker.sock --network=host \
       -v /sys/fs/cgroup:/sys/fs/cgroup -v `pwd`:/build:rw  centos:${OS_VERSION}   /usr/sbin/init
```

c-ney replied:
At least `rpmspec` doesn't like "incomplete" spec files. I'd think that something similar would happen in `yum-builddep`?

```
rpmspec -q --buildrequires --qf "[%{REQUIRENAME} ]" singularity.spec.in
error: line 28: Illegal char '@' (0x40) in: Version: @PACKAGE_VERSION@
error: query of specfile singularity.spec.in failed, can't parse
```

So going back to manual selection of the dependencies for the configure step?

olifre replied:
> So going back to manual selection of the dependencies for the configure step?

This has already gone wrong several times and will go wrong many times in the future, so I would strongly vote against it.

c-ney replied:
Then for now one could just

```
sed 's/@PACKAGE_VERSION@/0/' < singularity.spec.in > singularity_tmp.spec
```

as the spec file is otherwise left alone by autotools.

Concerning the unclean build folder "problem", maybe move the `cp -a . /tmp/$DIST/singularity` from `centos_run` to `before_install`?

jmstover commented on .travis/rpmbuild_test:
```diff
@@ -5,13 +5,13 @@
 make dist
 rpmbuild -ta *.tar.gz
-rpm -iv ~/rpmbuild/RPMS/*/*.rpm
+yum install -y ~/rpmbuild/RPMS/*/*.rpm
```
IMO, doing:  `yum localinstall -y ~/rpmbuild/RPMS/*/*.rpm` is better for an on disk location.

olifre replied:
Indeed! 
I would first wait for @c-ney who is basing his work on this PR if I understood correctly,
so I would not apply this change now since it may cause conflicts for him.
But this for sure should be changed to `localinstall` before the final version, thanks for the review! 👍

c-ney replied:
From what I get from the man page of yum, there is no difference between `yum install file.rpm` and `yum localinstall file.rpm` [1]. Also the latter is _kinda_ marked as deprecated. But thinking about it, I did not check this in centos 6. The version there might not be new enough.

> If the name is a file, then install works like localinstall.

> Note that the install command will do a local install, if given a filename. This command is maintained for legacy reasons only.

[1] http://man7.org/linux/man-pages/man8/yum.8.html
